apps/prairielearn/src/question-servers/freeform.js

```diff
@@ -1689,12 +1689,13 @@ module.exports = {
       );
   },
 
-  async fileAsync(filename, variant, question, course) {
+  async fileAsync(filename, variant, question, variant_course, question_course) {
     return instrumented('freeform.file', async (span) => {
       debug('file()');
       if (variant.broken) throw new Error('attemped to get a file for a broken variant');
 
-      const context = await module.exports.getContext(question, course);
+      const variantContext = await module.exports.getContext(question, variant_course);
+      const questionContext = await module.exports.getContext(question, question_course);
```

Comment:
I don't believe this is safe; look at what `getContext` actually does. Specifically, it will try to load chunks, which won't do anything sensible for the first call here where the question does not belong to `variant_course`. At any rate, AFAICT, `variantContext` is only used as an argument to `getCachedDataOrCompute`, but I think we actually _want_ `questionContext` to be used there, as it's the context that's actually used to compute the file (see the usage of `questionContext` inside the forth argument to `getCachedDataOrCompute`.

Reply:
The reason I did this is because I thought we would want the data created by `def file` to be stored in the context of the `variant_course`, that is, the course that is consuming the questions, not in the `question_course`, which is the course that shared the question. Is that right? Or do we want to create and cache the data under the course that shared the question? I just gave this a try and it seems to work fine, but I'm not sure if it could potentially cause problems in unexpected ways.

Reply:
No, I think we'd almost definitely want to use the question course + context for caching. That is, if the question course/context changes, we *must* invalidate the cache, and if the variant course/context changes, we *don't need to* invalidate the cache.

Reply:
Ok, sounds good. 

I guess I misunderstood what the cache was for and I was worried about course data leaking across courses.

I have made the change, hopefully this is good to go now.

Reply:
Yeah, the context w.r.t. the cache is just used as part of the cache key - if the context changes, we need the cache key to change too. If there's a specific concern you have with leaking data between courses, I'd love to hear it and hopefully resolve it.

Reply:
I guess there is no way data could be leaked between courses here in the sense of users seeing the wrong thing, but if you were attempting to quantify the amount of data use on server by course, then doing it this way might attribute the cached data to the wrong course. But I don't think we are doing that, are we? So it probably doesn't matter.

Reply:
We 100% are not doing that, yeah.